### DISCUSSION TITLE
Generating an access token while the OIDC Client is deprecated

### DISCUSSION THREAD

**Discussion by oussamabonnor1** [iTwin/itwinjs-core#3038] [Q&A] [answered]
We went for a machine-to-machine approach for the authentication using a [Token Server](https://github.com/calebmshafer/itwin-viewer-proxy/tree/main/token-server). 
This was a perfect way for us to seamlessly have our clients load their iModels inside an iTwin app without having to enter an email/password combo.
This was the old way of implementing this system into the viewer:

```jsx
<Viewer
  // ... (iModel related information)
  authConfig={{ oidcClient: MyTokenServerAuthClient.oidcClient }}
>
```

Since an `oidcClient` is no longer supported, what updates do we need to use for this new system ?

Is this relevant ? (Taken from https://github.com/iTwin/itwinjs-core/discussions/2158 ) :

![image](https://user-images.githubusercontent.com/17766221/149362254-a435d36e-0024-45b8-b80c-64e88e72eb4b.png)

**Comment by aruniverse** [comment, marked as answer]
Take a look at, https://github.com/iTwin/viewer/blob/next/NextVersion.md#authorization. The viewer expects the ViewerAuthorizationClient interface to be implemented, which the MyTokenServerAuthClient should already be. 

I believe the only change you need to make is 

```tsx
<Viewer
  // ... (iModel related information)
  authConfig={{ your instance of MyTokenServerAuthClient }}
>
```

**Reply by oussamabonnor1** [reply]
After using this interface in MyTokenServerAuthClient & making the necessary changes:

```tsx
interface ViewerAuthorizationClient {
  getAccessToken(): Promise<AccessToken>;
  readonly onAccessTokenChanged: BeEvent<(token: AccessToken) => void>;
}
```
An access token was generated using the server app but this error was thrown: 

![image](https://user-images.githubusercontent.com/17766221/149658306-d9e97c86-e0cd-447a-94b7-cb009b69b067.png)

I'm guessing that the code from the repo [itwin-viewer-proxy](https://github.com/calebmshafer/itwin-viewer-proxy/tree/main/token-server) (more precisely, this [code](https://github.com/calebmshafer/itwin-viewer-proxy/blob/main/token-server/src/AuthClient.ts)) is not generating iTwin V3 compatible tokens ?
If it helps, I can share here (or email) one of the tokens generated.

**Reply by hanan-commits-work** [reply]
Hi. @oussamabonnor1 , I had the same issue recently. 
Let me walk you through, how i resolved it.
1. Create your TokenServerAuth class that IMPLEMENTS AuthorizationClient
2. Create instance of your accessToken (make its type as string as the 3.0 accessToken is just a string containing "Bearer <token>") within your class
3. Create an instance of an AuthorizationClient
4. You can then use the functions from your old MyTokenAuthClient such as signIn, signOut, isAuthorized, etc
5. Make sure to implement generateToken and getAccessToken
6. In my case, my token string is located on a server, so I am simply making a fetch request in the generateToken and assigning it to this.accessToken
7. In your App.tsx, you can either create an instance of your TokenServerAuthClient anyway you want, I did it by useMemo. In your case it will be const object = useMemo ( () => new MyTokenServerAuthClient(), [] ); 
8. Then just assign this "object" to the AuthClient prop in your <Viewer>

I hope this helps. Let me know if you run into any other problems.

**Reply by oussamabonnor1** [reply]
@hanan-commits-work Thank you for the detailed answer.

I have already followed all the above mentioned steps except for:
- string as a type of the token, I used accessToken but I changed it now. I also added "Bearer "+token to the result.
- I switched from using an instance of TokenServerAuthClient using useState to useMemo.

Now the error switched from unsupported access token format to:

![image](https://user-images.githubusercontent.com/17766221/149748784-e0977fda-16b5-4265-86ac-30b9679913ae.png)

Since we already dealt with this, I thought solving it would require adding the digital twin management API scopes to the server app to include the new required scopes: `imodels:read` & `realitydata:read`

But, even with this.. The problem persists. Is there anything else I am missing ?

P.S: Before the error pops up, the `Your iModel is loading` message is shown briefly.

Just an FYI, this is using a service type app:

![image](https://user-images.githubusercontent.com/17766221/149749643-ff95f56d-461f-4134-85a0-ecf6bcb44f80.png)

**Reply by hanan-commits-work** [reply]
It could be that you havent defined user permissions in the bentley connect. I solved this issue by going into Projects -> <Project Name> -> iModel Manager (below in the iTwin services) -> opening options for the iModel used (youd see 3 dots with the iModel Name) -> set Model Access -> Granting the admin rights to the permission role of the service.. for my usecase, the service needed to have some of the admin rights, so i granted the full permissions of iModel to the "role of the admin"

**Reply by oussamabonnor1** [reply]
Already done, I believe the token server might need an update. Something that uses `@itwin/service-authorization`

**Reply by aruniverse** [reply]
I updated the sample to point to the 3.0 rc, https://github.com/calebmshafer/itwin-viewer-proxy. If you guys run into any issues please let me know.
